File: jpress-engine/src/main/java/by/ibragimov/jpress/generator/markup/AsciiDocEngine.java
package by.ibragimov.jpress.generator.markup;

import org.asciidoctor.Asciidoctor;
import org.asciidoctor.OptionsBuilder;
import org.springframework.stereotype.Service;

import java.io.File;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.FutureTask;
import java.util.concurrent.RunnableFuture;

/**
 * Converts input data to html markup.
 *
 * @author Ibragimov Ruslan
 */
@Service
public class AsciiDocEngine {

    private RunnableFuture<Asciidoctor> future = new FutureTask<>(Asciidoctor.Factory::create);

    public AsciiDocEngine() {
        future.run();
    }

    public AsciiDocConversionDto convert(File file) {
        try {
            Asciidoctor doc = future.get();
            return new AsciiDocConversionDto(doc.renderFile(file, OptionsBuilder.options()), doc.readDocumentHeader(file));
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }
}

File: kpress-engine/src/main/kotlin/by/heap/kpress/generator/markup/AsciiDocEngine.kt
package by.heap.kpress.generator.markup

import org.asciidoctor.Asciidoctor
import org.asciidoctor.OptionsBuilder
import java.io.File

/**
 * Converts input data to html markup.

 * @author Ruslan Ibragimov
 */
class AsciiDocEngine {
    private val asciidoctor = Asciidoctor.Factory.create()

    fun convert(file: File): AsciiDocConversionDto {
        val optionsBuilder = OptionsBuilder.options().inPlace(false)
        val fileContent = file.readText()
        val content = asciidoctor.convert(fileContent, optionsBuilder)

        return AsciiDocConversionDto(content, asciidoctor.readDocumentHeader(file))
    }
}
